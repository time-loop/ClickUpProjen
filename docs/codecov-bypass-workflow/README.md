# codecov-bypass

This is currently an explict opt-in feature. Eventually it will be a default enabled workflow with an explicit disable.

To add this to your project, you will need to

- edit your `.projenrc.ts`
- run `npx projen` to generate a bunch of files
- update your branch protection rules

## Edit your .projenrc.ts

```diff
 import { clickupCdk } from '@time-loop/clickup-projen';
 const project = new clickupCdk.ClickUpCdkTypeScriptApp({
   ...
+  codecovBypassOptionsConfig: {
+   githubAppId: '${{ vars.CODECOV_GITHUB_APP_ID }}',
+   githubAppPrivateKey: '${{ secrets.CODECOV_GITHUB_APP_PRIVATE_KEY }}'
+  },
```

## Generate new files

After the above change, run `npx projen` to generate the new files.

## Update Branch Protection Rules

Once the above changes are merged, you will need to update your branch protection rules:

1. Remove `codecod/patch`, and `codecov/patch` from the list of required checks.
2. Add `Code coverage increased` to the list of required checks.
