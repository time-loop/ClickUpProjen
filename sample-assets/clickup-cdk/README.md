[![codecov](https://codecov.io/gh/time-loop/WRITEME/branch/main/graph/badge.svg?token=WRITEME)](https://codecov.io/gh/time-loop/WRITEME)

# my-new-app-cdk
