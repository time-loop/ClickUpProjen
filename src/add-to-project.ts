import { typescript, YamlFile } from 'projen';

export module addToProjectWorkflow {
  export const RENOVATE_GITHUB_USERNAME = 'cu-infra-svc-git';

  export const DEFAULT_RENOVATE_PR_LABEL = 'renovate';
  export const GITHUB_PROJECT_NUMBER = '3';

  const defaultWorkflow = {
    name: 'add-to-project',
    on: {
      pull_request: {
        types: ['opened', 'labeled'],
      },
    },
    // Prevent renovate running multiple times in parallel due to several subsequent trigger events occurring
    concurrency: {
      group: "${{ github.workflow }}-${{ github.event_name == 'workflow_dispatch' && github.event_name || '' }}",
    },
    jobs: {
      'add-to-project': {
        'runs-on': 'ubuntu-latest',
        permissions: {
          contents: 'read',
        },
        steps: [
          {
            name: 'Checkout',
            uses: 'actions/checkout@v3',
          },
          {
            name: 'Add issue to project',
            uses: 'actions/add-to-project@v0.4.1',
            with: {
              // github project url
              'project-url': `https://github.com/orgs/time-loop/projects/${GITHUB_PROJECT_NUMBER}`,
              'github-token': 'x-access-token:${{ secrets.PROJEN_GITHUB_TOKEN }}',
              labeled: DEFAULT_RENOVATE_PR_LABEL,
            },
          },
        ],
      },
    },
  };

  export function addAddToProjectWorkflowYml(project: typescript.TypeScriptProject, override?: any): void {
    new YamlFile(project, '.github/workflows/add-to-project.yml', { obj: { ...defaultWorkflow, ...override } });
  }
}
