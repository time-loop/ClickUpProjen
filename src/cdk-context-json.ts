import { awscdk, JsonPatch, SampleFile } from 'projen';

export module cdkContextJson {
  export interface InjectionOptions {
    /**
     * The account number where the role the GH actions will assume lives.
     * This is usually the same account running cdk-pipelines,
     * but can be any account which has been configured with
     *
     * ```ts
     * cdk bootstrap --trust-for-lookup $this_account_number ...
     * ```
     */
    readonly lookupAccountId: string;
    /**
     * Which region should GH Auth into?
     * @default 'us-west-2'
     */
    readonly awsRegion?: string;
    /**
     * How long should the requested role be valid.
     * @default 900 15min seems pretty generous.
     */
    readonly roleDurationSeconds?: number;
  }

  /**
   * Options to support cdk.context.json management
   */
  export interface Options {
    /**
     * @default - do not create stack SampleCode for OIDC role
     */
    readonly createOidcRoleStack?: boolean;
    /**
     * You must configure this if you want to self-mutation cdk.context.json.
     *
     * @default - do not modify the build workflow to authorize against AWS.
     */
    readonly injectionOptions?: InjectionOptions;
  }

  export function injectAwsAuthIntoBuild(project: awscdk.AwsCdkTypeScriptApp, options: InjectionOptions): void {
    const build = project.tryFindObjectFile('.github/workflows/build.yml');
    build?.addOverride('jobs.build.permissions', { 'id-token': 'write' });
    build?.patch(
      JsonPatch.add('/jobs/build/steps/0', {
        name: 'Configure AWS Credentials',
        uses: 'aws-actions/configure-aws-credentials@v2',
        with: {
          'aws-region': options.awsRegion ?? 'us-west-2',
          'role-to-assume': `arn:aws:iam::${options.lookupAccountId}:role/${project.name}-github-actions-role`,
          'role-duration-seconds': options.roleDurationSeconds ?? 900,
        },
      }),
    );
  }

  export function addOidcRoleStack(project: awscdk.AwsCdkTypeScriptApp): void {
    new SampleFile(project, `${project.srcdir}/github-actions-oidc-cdk-context-lookup-role.ts`, {
      contents: `import { core } from '@time-loop/cdk-library';
import { aws_iam, Stage } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { Namer } from 'multi-convention-namer';

export class GitHubActionsOIDCCdkContextLookupRole extends core.Stack {
  static asStage(scope: Construct, stageName: string, stageProps: core.StageProps): Stage {
    return new (class extends Stage {
      constructor() {
        super(scope, stageName, stageProps);
        new GitHubActionsOIDCCdkContextLookupRole(this, stageProps);
      }
    })();
  }
  constructor(scope: Construct, props: core.StackProps) {
    const projectName = '${project.name}';
    // NOTE: you should never deploy this anywhere but the cdkPipelines account
    // And the GitHubActionsOIDCPermissions class for cdk-diff should never be deployed
    // in the cdkPipelines account. Name collision is intentional here.
    let id = new Namer([...projectName.split('-'), 'github', 'actions']);
    super(scope, id, props);

    const githubActionsRoleName = id.addSuffix(['role']).kebab;
    const githubActionsRole = new aws_iam.Role(this, githubActionsRoleName, {
      roleName: githubActionsRoleName,
      assumedBy: new aws_iam.FederatedPrincipal(
        \`arn:aws:iam::\${this.account}:oidc-provider/token.actions.githubusercontent.com\`,
        {
          StringEquals: {
            'token.actions.githubusercontent.com:aud': 'sts.amazonaws.com',
          },
          StringLike: {
            'token.actions.githubusercontent.com:sub': \`repo:time-loop/\${projectName}:*\`,
          },
        },
        'sts:AssumeRoleWithWebIdentity',
      ),
    });

    const githubActionsPolicyName = id.addSuffix(['policy']).kebab;
    const githubActionsPolicy = new aws_iam.Policy(this, githubActionsPolicyName, {
      policyName: githubActionsPolicyName,
      statements: [
        new aws_iam.PolicyStatement({
          actions: [ 'sts:AssumeRole' ],
          resources: [ 'arn:aws:iam::*:role/cdk-*lookup-role*' ],
        }),
      ],
    });

    // Attach IAM policy to IAM role
    githubActionsPolicy.attachToRole(githubActionsRole);
  }
}
`,
    });
  }
}
