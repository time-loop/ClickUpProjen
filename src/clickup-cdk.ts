import * as path from 'path';
import { awscdk, Component, SampleDir, SampleReadme } from 'projen';
import merge from 'ts-deepmerge';

import { clickupTs } from './clickup-ts';
import { codecov } from './codecov';

export module clickupCdk {
  export const deps = [
    ...clickupTs.deps,
    '@time-loop/cdk-library',
    'cdk-constants',
    'cdk-iam-floyd',
    'chalk',
    'multi-convention-namer',
  ];
  export const defaults = {
    deps,
    jestOptions: {
      jestConfig: {
        coveragePathIgnorePatterns: ['/node_modules/', '/src/main.ts'],
      },
    },
    releaseToNpm: false,
  };

  export interface ClickUpCdkConstructLibraryOptions extends awscdk.AwsCdkConstructLibraryOptions {}

  /**
   * ClickUp standardized CDK Construct Library.
   *
   * Note: disgusting hack to achieve "defaults" in the constructor
   * leverages "empty string is falsy" behavior of TS.
   * I am not proud of this.
   * It's better than cloning the interface since projen revs pretty fast.
   * Marginally.
   */
  export class ClickUpCdkConstructLibrary extends awscdk.AwsCdkConstructLibrary {
    constructor(options: ClickUpCdkConstructLibraryOptions) {
      // JSII means I can't Omit and then re-implement the following as optional. So...
      const authorName = options.author || clickupTs.defaults.authorName;
      const authorAddress = options.authorAddress || clickupTs.defaults.authorAddress;
      const githubOwner = `${process.env.GITHUB_OWNER ?? 'time-loop'}`;
      // Theoretically we should be able to just take a default here, but for some reason this is required.
      const repositoryUrl = options.repositoryUrl || `https://github.com/${githubOwner}/${options.name}.git`;
      super(merge(clickupTs.defaults, options, { authorName, authorAddress, repositoryUrl }));
      codecov.addCodeCovYml(this);
      codecov.addCodeCovOnRelease(this);
    }
  }

  export interface ClickUpCdkTypeScriptAppOptions extends awscdk.AwsCdkTypeScriptAppOptions {}

  /**
   * ClickUp standardized CDK TypeScript App
   *
   * Includes:
   * - default author information
   * - default proprietary license
   * - default release build configuration
   * - default linting and codecov configuration
   * - default minNodeVersion: '14.17.0'
   * - default deps and devDeps (you can add your own, but the base will always be present)
   */
  export class ClickUpCdkTypeScriptApp extends awscdk.AwsCdkTypeScriptApp {
    constructor(options: ClickUpCdkTypeScriptAppOptions) {
      super(merge(clickupTs.defaults, defaults, options, { sampleCode: false }));
      new SampleCode(this);
      new SampleReadme(this, {
        contents: `[![codecov](https://codecov.io/gh/time-loop/WRITEME/branch/main/graph/badge.svg?token=WRITEME)](https://codecov.io/gh/time-loop/WRITEME)

        # my-new-app-cdk
        `,
      });
      codecov.addCodeCovYml(this);
      codecov.addCodeCovOnRelease(this);
    }
  }

  class SampleCode extends Component {
    constructor(project: ClickUpCdkTypeScriptApp) {
      super(project);

      // src files
      new SampleDir(project, project.srcdir, {
        files: {
          'main.ts': `import { core } from '@time-loop/cdk-library';
import { App } from 'aws-cdk-lib';
import 'colors';
import { Namer } from 'multi-convention-namer';

import { WidgetStack } from './widget';

const app = new App();
const env = process.env.AWS_PROFILE || (process.env.CI ? 'usDev' : '');
const region = process.env.AWS_REGION || (process.env.CI ? 'us-east-1' : undefined);
if (!env) {
  console.log('You should probably set AWS_PROFILE before using this.'.yellow);
}
const namedEnvFactory = core.Environment.findByName(env);
const namedEnv = namedEnvFactory(region);

console.log(\`Deploying to \${JSON.stringify(namedEnv.name).blue} in \${JSON.stringify(namedEnv.region).blue}.\`);

const commonProps = {
  businessUnit: core.BusinessUnit.PRODUCT,
  clickUpEnvironment: core.ClickUpEnvironment.PRODUCTION,
  clickUpRole: core.ClickUpRole.APP,
  confidentiality: core.Confidentiality.PUBLIC,
  namedEnv,
};
new WidgetStack(app, new Namer(['my', 'cool', 'widget']), {
  ...commonProps,
  managedPolicyName: 'yarf',
});

app.synth();
`,
          'widget.ts': `
import { core } from '@time-loop/cdk-library';
import { aws_iam, aws_kms, RemovalPolicy } from 'aws-cdk-lib';
import * as statement from 'cdk-iam-floyd';
import { Construct } from 'constructs';
import { Namer } from 'multi-convention-namer';

export interface WidgetProps {
  /**
   * The name to assign the widget policy.
   * @default - have CDK generate a unique name
   */
  readonly managedPolicyName?: string;
}

/**
 * Useful comment describing the Widget construct.
 */
export class Widget extends Construct {
  // Expose the policy for use by another Construct in this Stack.
  // WARNING: cdk absolutely will let you pass objects between stacks.
  // This will generate CfnOutputs that will in turn create tight coupling between stacks.
  // This is almost never a Good Idea.
  // Instead, consider using \`core.Param.put\` to put the ARN
  // and then use myThingy.fromThingyArn() to instantiate a Interface object.
  readonly policy: aws_iam.ManagedPolicy;

  constructor(scope: Construct, id: Namer, props: WidgetProps) {
    super(scope, id.pascal);

    const key = new aws_kms.Key(this, 'Key', {
      removalPolicy: RemovalPolicy.DESTROY,
    });
    // Save the key ARN using SSM, for use by another Stack
    core.Param.put(key, 'keyArn', { rootId: 'ThisAppName' });

    this.policy = new aws_iam.ManagedPolicy(this, 'Policy', {
      managedPolicyName: props.managedPolicyName,
    });
    this.policy.addStatements(
      // Reference the ARN of a locally created thingy.
      // Note that the \`allow()\` is implicit.
      new statement.Kms({ sid: 'DescriptiveName' }).to('Decrypt*').on(key.keyArn),
      // Have \`cdk-iam-floyd\` construct the ARN for you
      new statement.Cloudfront({ sid: 'GrantInvalidation' }).toCreateInvalidation().onDistribution(
        // Pull the distributionId from an SSM parameter
        core.Param.get(this, 'distributionId', {
          rootId: 'SomeOtherApp',
          stackId: 'SomeOtherStack',
          constructId: 'FooBar',
        }),
      ),
    );
  }
}

export interface WidgetStackProps extends core.StackProps, WidgetProps {}

export class WidgetStack extends core.Stack {
  constructor(scope: Construct, id: Namer, props: WidgetStackProps) {
    super(scope, id, props);

    const widget = new Widget(this, id, props);

    // Example of passing the widget.policy to another construct in the same stack.
    // \`\`\`
    // new FrozzleBop(this, id, { policy: widget.policy });
    // \`\`\`
    // placeholder to keep lint happy
    console.log(widget.policy.managedPolicyArn);
  }
}
`,
        },
      });

      // test files
      new SampleDir(project, project.testdir, {
        files: {
          'widget.test.ts': `import { core } from '@time-loop/cdk-library';
import { App, assertions } from 'aws-cdk-lib';
import { Namer } from 'multi-convention-namer';
import { WidgetStack } from '../src/widget';

// Minimum props required by @time-loop/cdk-library/core.StackProps
const commonProps = {
  businessUnit: core.BusinessUnit.PRODUCT,
  clickUpEnvironment: core.ClickUpEnvironment.PRODUCTION,
  clickUpRole: core.ClickUpRole.APP,
  confidentiality: core.Confidentiality.PUBLIC,
  namedEnv: core.Environment.usDev('us-west-2'),
};

describe('Widget', () => {
  describe('default', () => {
    // These are resources shared by all the tests in this context.
    const app = new App();
    const stack = new WidgetStack(app, new Namer(['test']), commonProps);
    const template = assertions.Template.fromStack(stack);

    // Tests are super thin, consisting of just an assertion.
    test('snapshot', () => {
      // This is the fastest, most superficial test you can write.
      // It's better than nothing. But kinda fragile.
      expect(template.toJSON()).toMatchSnapshot();
    });

    // Demonstrate super-cool matcher stuff
    test('policy should reference key', () => {
      template.hasResourceProperties('AWS::IAM::ManagedPolicy', {
        PolicyDocument: {
          // The statement array must contain the following array.
          Statement: assertions.Match.arrayWith([
            assertions.Match.objectLike({
              // The array must contain an object with at least the following key / value.
              Resource: {
                'Fn::GetAtt': [
                  assertions.Match.anyValue(), // TODO: figure out how to actually reference the generated Kms key
                  'Arn',
                ],
              },
            }),
          ]),
        },
      });
    });
  });

  describe('options', () => {
    // Here we aren't sharing setup code...
    test('managedPolicyName', () => {
      // ...because each test is exercising a specific part of the functionality.
      // Which means that setups are slightly different and we can't re-use things.
      const app = new App();
      const stack = new WidgetStack(app, new Namer(['test']), {
        ...commonProps,
        managedPolicyName: 'fakeName',
      });
      const template = assertions.Template.fromStack(stack);

      template.hasResourceProperties('AWS::IAM::ManagedPolicy', {
        ManagedPolicyName: 'fakeName',
      });
    });
  });
});
`,
        },
      });
      new SampleDir(project, path.join(project.testdir, '__snapshots__'), {
        files: {
          'widget.test.ts.snap': `// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[\`Widget default snapshot 1\`] = \`
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "SsmParameterValueSomeOtherAppSomeOtherStackFooBardistributionIdC96584B6F00A464EAD1953AFF4B05118Parameter": Object {
      "Default": "/SomeOtherApp/SomeOtherStack/FooBar/distributionId",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "TestKey4BC8CF8E": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "KeyPolicy": Object {
          "Statement": Array [
            Object {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::469006742758:root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": Array [
          Object {
            "Key": "Business Unit",
            "Value": "product",
          },
          Object {
            "Key": "Confidentiality",
            "Value": "public",
          },
          Object {
            "Key": "DATADOG",
            "Value": "true",
          },
          Object {
            "Key": "Environment",
            "Value": "production",
          },
          Object {
            "Key": "Role",
            "Value": "app",
          },
          Object {
            "Key": "Service",
            "Value": "Test",
          },
        ],
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Delete",
    },
    "TestKeyParameterThisAppNameTestKeyKeyArn13E268B9": Object {
      "Properties": Object {
        "Name": "/ThisAppName/Test/Key/keyArn",
        "Tags": Object {
          "Business Unit": "product",
          "Confidentiality": "public",
          "DATADOG": "true",
          "Environment": "production",
          "Role": "app",
          "Service": "Test",
        },
        "Type": "String",
        "Value": Object {
          "Fn::GetAtt": Array [
            "TestKey4BC8CF8E",
            "Arn",
          ],
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "TestPolicyAC2892FA": Object {
      "Properties": Object {
        "Description": "",
        "Path": "/",
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "kms:Decrypt*",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "TestKey4BC8CF8E",
                  "Arn",
                ],
              },
              "Sid": "DescriptiveName",
            },
            Object {
              "Action": "cloudfront:CreateInvalidation",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:cloudfront::*:distribution/",
                    Object {
                      "Ref": "SsmParameterValueSomeOtherAppSomeOtherStackFooBardistributionIdC96584B6F00A464EAD1953AFF4B05118Parameter",
                    },
                  ],
                ],
              },
              "Sid": "GrantInvalidation",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::ManagedPolicy",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
\`;
`,
        },
      });
    }
  }
}
