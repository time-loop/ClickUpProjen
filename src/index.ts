export * from './clickup-cdk';
export * from './clickup-ts';
export * from './datadog';
export * from './slack-alert';
export * from './renovate-workflow';
export * from './cdk-diff-workflow';
export * from './add-to-project';
