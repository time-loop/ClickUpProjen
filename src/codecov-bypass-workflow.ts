import { YamlFile } from 'projen';
import { clickupCdk } from './clickup-cdk';

const CODECOV_GITHUB_APP_ID = 254;

export module codecovBypassWorkflow {
  function createCodecovBypassWorkflow(options?: CodecovBypassOptionsConfig) {
    const defaultWorkflow = {
      name: options?.workflowName || 'Code coverage increased',
      on: {
        check_suite: {
          types: ['completed'],
        },
        pull_request: {
          types: ['labeled', 'unlabeled'],
        },
      },
      permissions: {},
      jobs: {
        'code-coverage-increased': {
          'runs-on': 'ubuntu-latest',
          if: `github.event_name != \'check_suite\' || github.event.check_suite.app.id == ${CODECOV_GITHUB_APP_ID}`,
          steps: [
            {
              name: 'Check codecov results',
              uses: 'time-loop/github-actions/dist/wrap-check-suite@wrap-check-suite+0.3.2',
              env: {
                FORCE_COLOR: 3,
              },
              with: {
                'github-app-id': options?.githubAppId || '${{ vars.CODECOV_GITHUB_APP_ID }}',
                'github-private-key': options?.githubAppPrivateKey || '${{ secrets.CODECOV_GITHUB_APP_PRIVATE_KEY }}',
                'skip-label': options?.skipLabel || 'code coverage not required',
                'check-name': options?.checkName || 'Code coverage increased',
                'check-suite-app-id': CODECOV_GITHUB_APP_ID,
                'check-suite-check-names': options?.checkSuiteCheckNames?.join(',') || 'codecov/patch,codecov/project',
                'details-url':
                  options?.detailsUrl || 'https://app.codecov.io/gh/${{ github.repository }}/pull/<% prNumber %>',
              },
            },
          ],
        },
      },
    };

    return defaultWorkflow;
  }

  export interface CodecovBypassOptionsConfig {
    /**
     * GitHub App ID
     * @default '${{ vars.CODECOV_GITHUB_APP_ID }}'
     */
    readonly githubAppId?: string;

    /**
     * GitHub App Private Key
     * @default '${{ secrets.CODECOV_GITHUB_APP_PRIVATE_KEY }}'
     */
    readonly githubAppPrivateKey?: string;

    /**
     * Skip creating (using) the workflow
     * @default false
     */
    readonly disabled?: boolean;

    /**
     * Workflow Name
     * @default 'Code coverage increased'
     */
    readonly workflowName?: string;

    /**
     * Skip Label
     * @default 'code coverage not required'
     */
    readonly skipLabel?: string;

    /**
     * Check Name
     * @default 'Code coverage increased'
     */
    readonly checkName?: string;

    /**
     * Check Suite Check Names
     * @default ['codecov/patch', 'codecov/project']
     */
    readonly checkSuiteCheckNames?: string[];

    /**
     * Details URL
     * @default 'https://app.codecov.io/gh/${{ github.repository }}/pull/<% prNumber %>'
     */
    readonly detailsUrl?: string;
  }

  export function getCodecovBypassOptions(options?: CodecovBypassOptionsConfig) {
    return options;
  }

  export function addCodecovBypassWorkflowYml(
    project: clickupCdk.ClickUpCdkTypeScriptApp,
    options?: CodecovBypassOptionsConfig,
    override?: any,
  ): void {
    if (options?.disabled) {
      return;
    }
    new YamlFile(project, '.github/workflows/codecov-bypass.yml', {
      obj: { ...createCodecovBypassWorkflow(options), ...override },
    });
  }
}
