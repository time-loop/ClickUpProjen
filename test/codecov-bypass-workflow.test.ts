import { Testing } from 'projen';

import { requiredParams } from './requiredParams';
import { clickupCdk } from '../src/clickup-cdk';
import { codecovBypassWorkflow } from '../src/codecov-bypass-workflow';

describe('addCodecovBypassWorkflowYml - codecov-bypass .yml file added', () => {
  test('all default options', () => {
    const project = new clickupCdk.ClickUpCdkTypeScriptApp(requiredParams);
    codecovBypassWorkflow.addCodecovBypassWorkflowYml(project);
    const synth = Testing.synth(project);
    expect(synth['.github/workflows/codecov-bypass.yml']).toMatchSnapshot();
  });

  test('all options provided', () => {
    const project = new clickupCdk.ClickUpCdkTypeScriptApp(requiredParams);
    codecovBypassWorkflow.addCodecovBypassWorkflowYml(project, {
      workflowName: 'workflowName',
      githubAppId: '${{ vars.GH_APP_ID_VAR_NAME }}',
      githubAppPrivateKey: 'r${{ secrets.GH_APP_ID_PRIVATE_KEY_SECRET_NAME }}',
      skipLabel: 'skipLabel',
      checkName: 'checkName',
      checkSuiteCheckNames: ['checkSuiteCheckName1', 'checkSuiteCheckName2'],
      detailsUrl: 'some url',
    });
    const synth = Testing.synth(project);
    expect(synth['.github/workflows/codecov-bypass.yml']).toMatchSnapshot();
  });

  test('disabled', () => {
    const project = new clickupCdk.ClickUpCdkTypeScriptApp(requiredParams);
    codecovBypassWorkflow.addCodecovBypassWorkflowYml(project, {
      disabled: true,
    });
    const synth = Testing.synth(project);
    expect(synth['.github/workflows/codecov-bypass.yml']).toBeUndefined();
  });
});
