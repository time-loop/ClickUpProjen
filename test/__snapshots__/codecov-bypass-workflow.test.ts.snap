// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addCodecovBypassWorkflowYml - codecov-bypass .yml file added all default options 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: Code coverage increased
on:
  check_suite:
    types:
      - completed
  pull_request:
    types:
      - labeled
      - unlabeled
permissions: {}
jobs:
  code-coverage-increased:
    runs-on: ubuntu-latest
    if: github.event_name != 'check_suite' || github.event.check_suite.app.id == 254
    steps:
      - name: Check codecov results
        uses: time-loop/github-actions/dist/wrap-check-suite@wrap-check-suite+0.3.2
        env:
          FORCE_COLOR: 3
        with:
          github-app-id: \${{ vars.CODECOV_GITHUB_APP_ID }}
          github-private-key: \${{ secrets.CODECOV_GITHUB_APP_PRIVATE_KEY }}
          skip-label: code coverage not required
          check-name: Code coverage increased
          check-suite-app-id: 254
          check-suite-check-names: codecov/patch,codecov/project
          details-url: https://app.codecov.io/gh/\${{ github.repository }}/pull/<% prNumber %>
"
`;

exports[`addCodecovBypassWorkflowYml - codecov-bypass .yml file added all options provided 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: workflowName
on:
  check_suite:
    types:
      - completed
  pull_request:
    types:
      - labeled
      - unlabeled
permissions: {}
jobs:
  code-coverage-increased:
    runs-on: ubuntu-latest
    if: github.event_name != 'check_suite' || github.event.check_suite.app.id == 254
    steps:
      - name: Check codecov results
        uses: time-loop/github-actions/dist/wrap-check-suite@wrap-check-suite+0.3.2
        env:
          FORCE_COLOR: 3
        with:
          github-app-id: \${{ vars.GH_APP_ID_VAR_NAME }}
          github-private-key: r\${{ secrets.GH_APP_ID_PRIVATE_KEY_SECRET_NAME }}
          skip-label: skipLabel
          check-name: checkName
          check-suite-app-id: 254
          check-suite-check-names: checkSuiteCheckName1,checkSuiteCheckName2
          details-url: some url
"
`;
